File: 1002-maximum-width-ramp/maximum-width-ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        
        int n=nums.size();
        vector<int>indices(n,0);
        for(int i=0;i<n;i++) indices[i]=i;

        sort(indices.begin(),indices.end(),[&](int i,int j){
            return nums[i]!=nums[j]?nums[i]<nums[j]:i<j;
        });

        int ramp=0;
        int min_ind=indices[0];
        for(int i=1;i<n;i++){
            min_ind=min(min_ind,indices[i]);
            ramp=max(ramp,indices[i]-min_ind);
        }
        return ramp;
        }
};

File: 1002-maximum-width-ramp/maximum-width-ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int n=nums.length;
        int []suffix = new int[n];
        suffix[n-1]=nums[n-1];
        for(int i=n-2;i>=0;i--){

            suffix[i]=Math.max(suffix[i+1],nums[i]);
        }

        int ramp=0;
        int j=0;
        for(int i=0;i<n;i++){

            while(j<n && nums[i]<=suffix[j]){
                j++;
            }
            j--;
            ramp=Math.max(j-i,ramp);
        }
        return ramp;
    }
}